definitions/npm/async_v1.x.x/flow_>=v0.28.x/async_v1.x.x.js

```diff
@@ -0,0 +1,77 @@
+// @flow
+
+declare module 'async' {
+  declare type Collection<TItem> = Iterable<TItem> | { [key: string]: TItem };
+
+  declare type IterateeCallback<TError, TResult> = (err?: TError, result?: TResult) => void;
+  declare type Callback<TError, TResult> = (err?: TError, result?: TResult) => mixed;
+
+  // Collections
+  declare function each<TItem, TError>(
+    collection: Collection<TItem>,
+    iteratee: (item: TItem, callback: IterateeCallback<TError, void>) => mixed,
+    callback?: Callback<TError, void>
+  ): void;
+  declare function forEachOf<TItem, TError>(
+    collection: Collection<TItem>,
+    iteratee: (value: TItem, key: string | number, callback: IterateeCallback<TError, void>) => mixed,
+    callback?: Callback<TError, void>
+  ): void;
+  declare function map<TItem, TResult, TError>(
+    collection: Collection<TItem>,
+    iteratee: (value: TItem, callback: IterateeCallback<TError, TResult>) => mixed,
+    callback?: Callback<TError, Array<TResult>>
```

Comment:
> I tried declaring two separate methods, but then flow can't decide which to use.

cc @avikchaudhuri -- This seems like something we should try to improve with the latest intersections changes.

I think the right way to type this is with overloads like you tried, but Flow is a bit infantile with this right now...so it's not great for libdef consumers. For now I'd say just define the `error` param on the callback as `null | TError` and maybe write a TODO to revisit this once Flow has made some progress with deciding how to select overrides (hopefully in a near-future version?).

This will require that people always check for `null` in their callback, but at least that's still safe.